### PR TITLE
contrib/dimfeld/httptreemux/v5: add support for httptreemux

### LINKED ISSUE #1520: contrib/dimfeld/httptreemux/v5: Add support for dimfeld/httptreemux
We use [dimfeld/httptreemux](https://github.com/dimfeld/httptreemux) as our router. Unfortunately there is no direct Datadog tracing support for `dimfeld/httptreemux`.

I would like to propose adding support for Datadog tracing to the `dimfeld/httptreemux` project. This would allow us to easily instrument our router based on `dimfeld/httptreemux`, as well as other folks that also use this router.

The process should be relatively straightforward and similar to what was done for [Julien Schmidt's httprouter](https://www.github.com/julienschmidt/httprouter) in [contrib/julienschmidt/httprouter](https://github.com/DataDog/dd-trace-go/tree/main/contrib/julienschmidt/httprouter).

The only major difference between `julienschmidt/httprouter` and `dimfeld/httptreemux` in terms of affecting Datadog tracing is the way in which they deal with URL parameters, i.e.:
- `julienschmidt/httprouter` uses a custom `[]Param` type where `Param` is a custom type with a string key value pair (see [julienschmidt/httprouter/router.go#L92](https://github.com/julienschmidt/httprouter/blob/master/router.go#L92))
- Where `dimfeld/httptreemux` uses a `map[string]string`.

~The only blocker to adding support for Datadog tracing to `dimfeld/httptreemux` is the fact that the `httptreemux.LookupResult` type contains a private `params` field that stores the path parameters. These need to be accessible in order to use the path params to build the resource name. I have created an issue (https://github.com/dimfeld/httptreemux/issues/89) and PR (https://github.com/dimfeld/httptreemux/pull/90) to make the necessary changes.~ [Resolved](https://github.com/DataDog/dd-trace-go/issues/1520#issuecomment-1380944411).

### PR DESCRIPTION
**Pull request by devillexio**
Adding support to trace the dimfeld/httptreemux/v5 router package. The changes are almost identical to those that were required to instrument julienschmidt/httprouter, since httptreemux is based on httprouter.

Resolves #1520.

### CODE REVIEW

**Comment by devillexio** [comment]
Hi friendly folks! :wave:

Just wanted to check if there is anything needed from us for the time being? Looking forward to your initial review! :raised_hands:

**Comment by ajgajg1134** [comment]
Hello! The team does periodic proposal reviews where we go through the github issues marked with `proposal` (Which I've just added to your linked issue #1520). We'll leave comments there once we've evaluated it. Thanks!

**Comment by laughingman-hass** [comment]
I would really love to see this PR merged

**Comment by ajgajg1134** [comment]
We just looked at the proposal and accepted it! We'll do a full review of this PR soon :)

File: contrib/dimfeld/httptreemux/v5/httptreemux.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+func New(opts ...RouterOption) *Router {
+	cfg := new(routerConfig)
+	defaults(cfg)
+	for _, fn := range opts {
+		fn(cfg)
+	}
+	if !math.IsNaN(cfg.analyticsRate) {
+		cfg.spanOpts = append(cfg.spanOpts, tracer.Tag(ext.EventSampleRate, cfg.analyticsRate))
+	}
+	cfg.spanOpts = append(cfg.spanOpts, tracer.Measured())
```
Comment (katiehockman) [inline contrib/dimfeld/httptreemux/v5/httptreemux.go:38]: A few more `spanOpts` to add:

```
cfg.spanOpts = append(cfg.spanOpts, tracer.Tag(ext.SpanKind, ext.SpanKindServer))
cfg.spanOpts = append(cfg.spanOpts, tracer.Tag(ext.Component, "dimfeld/httptreemux.v5"))
```
https://github.com/DataDog/dd-trace-go/tree/main/contrib#usage

File: contrib/dimfeld/httptreemux/v5/contextmux_test.go
Comment (katiehockman) [inline contrib/dimfeld/httptreemux/v5/contextmux_test.go:1]: All of this code should be in `contrib/dimfeld/httptreemux.v5/` per our naming conventions: https://github.com/DataDog/dd-trace-go/tree/main/contrib#usage

File: contrib/dimfeld/httptreemux/v5/httptreemux.go (hunk 1 above)
Comment (katiehockman) [reply contrib/dimfeld/httptreemux/v5/httptreemux.go:38]: Same thing for `ContextRouter`

File: contrib/dimfeld/httptreemux/v5/httptreemux.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+	log.Debug("contrib/dimfeld/httptreemux/v5: Configuring Router: %#v", cfg)
```
Comment (katiehockman) [inline contrib/dimfeld/httptreemux/v5/httptreemux.go:73]:
```suggestion
	log.Debug("contrib/dimfeld/httptreemux/v5: Configuring ContextRouter: %#v", cfg)
```

File: contrib/dimfeld/httptreemux/v5/option.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+	"github.com/dimfeld/httptreemux/v5"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/globalconfig"
+)
+
+type routerConfig struct {
+	serviceName   string
+	spanOpts      []ddtrace.StartSpanOption
+	analyticsRate float64
```
Comment (katiehockman) [inline contrib/dimfeld/httptreemux/v5/option.go:22]: We actually don't need this anymore, as this is a deprecated feature. No need to remove it from other library integrations right now, but also no need to introduce this in new integrations. So you can just remove this variable and all usages of it in this package.

File: contrib/dimfeld/httptreemux/v5/option.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+func WithAnalyticsRate(rate float64) RouterOption {
+	return func(cfg *routerConfig) {
+		if rate >= 0.0 && rate <= 1.0 {
+			cfg.analyticsRate = rate
+		} else {
+			cfg.analyticsRate = math.NaN()
+		}
+	}
+}
+
```
Comment (katiehockman) [inline contrib/dimfeld/httptreemux/v5/option.go:78]: All of this can be deleted

File: contrib/dimfeld/httptreemux/v5/httptreemux_test.go [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+		mt := mocktracer.Start()
+		defer mt.Stop()
+
+		rate := globalconfig.AnalyticsRate()
+		defer globalconfig.SetAnalyticsRate(rate)
+		globalconfig.SetAnalyticsRate(0.4)
+
+		assertRate(t, mt, 0.23, WithAnalyticsRate(0.23))
+	})
+}
```
Comment (katiehockman) [inline contrib/dimfeld/httptreemux/v5/httptreemux_test.go:159]: This can be removed

**Review by katiehockman — COMMENTED** [review]
Would you mind rebasing your branch? Thanks for your contribution! Just a few small comments, but overall this looks good.

**Comment by devillexio** [comment]
Thank you for the review and feedback @katiehockman! 🙌

I will work through the requested changes and submit an update as soon as possible.

File: contrib/dimfeld/httptreemux/v5/httptreemux_test.go (hunk 5 above)
Comment (devillexio) [reply contrib/dimfeld/httptreemux/v5/httptreemux_test.go:159]: - [ ] remove TestAnalyticsSettings

File: contrib/dimfeld/httptreemux/v5/option.go (hunk 3 above)
Comment (devillexio) [reply contrib/dimfeld/httptreemux/v5/option.go:22]: - [ ] remove `analyticsRate` and usages

File: contrib/dimfeld/httptreemux/v5/option.go (hunk 4 above)
Comment (devillexio) [reply contrib/dimfeld/httptreemux/v5/option.go:78]: - [ ] remove `WithAnalyticsRate` and usages

File: contrib/dimfeld/httptreemux/v5/contextmux_test.go
Comment (devillexio) [reply contrib/dimfeld/httptreemux/v5/contextmux_test.go:1]: - [ ] move `contrib/dimfeld/httptreemux/v5/` to `contrib/dimfeld/httptreemux.v5/`

File: contrib/dimfeld/httptreemux/v5/httptreemux.go (hunk 1 above)
Comment (devillexio) [reply contrib/dimfeld/httptreemux/v5/httptreemux.go:38]: - [ ] add missing `spanOpts` to `Router`
- [ ] add missing `spanOpts` to `ContextRouter`

**Comment by devillexio** [comment]
Thanks again for the review and feedback @katiehockman. 🙌  I believe I have addressed all the points of feedback. Please take a look when you have a chance. 🙏

**Review by katiehockman — APPROVED** [review]
Thank you!!

**Comment by katiehockman** [comment]
Looks like there are some lint errors, but if you could fix those @devillexio then we should be good to merge!

**Comment by katiehockman** [comment]
Merged, thanks again @devillexio !!! 🎉

**Pull request merged**